### ISSUE TITLE
Add Support for Java 21 Record Pattern : New Check unneededNullCheckWithInstanceof

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#14945]
child of #14890
 Previously proposed in #6943 

when using pattern matching with instance of,  the null-check is redundant as instanceof operator implies non-nullity. It will be good if we have a check that violates this redundant condition to lead to a more concise and clear code

**Config:**

```XML
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="unneededNullCheckWithInstanceof"/>
    </module>
</module>
```
---

**Example 1**

```java
public class Test {
   void test(Object obj) {
        if (obj != null && obj instanceof String) {   // violation
            System.out.println("string");
        }
    }
}
```
**Expected Output**

```
$ java -jar checkstyle-10.19.0-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] D:\Test.java:3:17: Unneeded nullity check with Instanceof as instanceof operator implies non-nullity
Audit done.
Checkstyle ends with 1 errors.
```

the check should violate the null check here because Condition `obj != null` is covered by subsequent
 condition `obj instanceof String` `

---

**Example 2**

```java
class Test {
    record ColoredPoint(int x, int y, String c) { }
    void test(Object obj) {
        if (obj instanceof ColoredPoint(_,_,_) && obj != null) { // violation
            System.out.println("string");
        }
    }
}
```
**Expected Output**

```
$ java -jar checkstyle-10.19.0-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] D:\Test.java:4:55: Unneeded nullity check with Instanceof as instanceof operator implies non-nullity
Audit done.
Checkstyle ends with 1 errors.
```

the check should violate the null check here because Condition `obj != null` is always 'true' when reached

**Comment by rnveach** [comment]
I am fine with the issue.

Issue should expand on some cases we will not support if we know so everything is being brought forward.
Previously said we would not support: `if ((myObj != null || abc) && myObj instanceof SomeObject) {}`.
Previously said we would support: `if ((myObj != null) && (myObj instanceof SomeObject)) {}`
@mahfouz72 You don't always have to split out all examples in their own Java files, however you don't need to always squash them in. Whatever makes it easier to get the general issue idea out there.

Check names should not start with a lowercase. They should be camel case, with an upper first letter.

Also, I think `Of` should also be camel (capital o). I won't really block this if others disagree as Java has it all lowercase, but it slightly seems odd to me.

https://checkstyle.org/checks.html
I think we should follow similar patterns we have for other check names. So either this should be an "Avoid" check or an "Unnecessary" check or a "No" check. To me this seems like an "Unnecessary" check. `instanceof` is already null safe in this case.

No other issues with the rest of the check name.

**Comment by romani** [comment]
I like NullCheckWithInstanceof name.
Do you see problem with it?

**Comment by mahfouz72** [comment]
> Previously said we would support:
> 
> ```
> if ((myObj != null) && (myObj instanceof SomeObject)) {}
> if (myObj != null) {
>   if (myObj instanceof SomeObject) {
>   }
> }
> ```

I have a comment on this example consider this variation from it:
```
 if (myObj != null) { 
    if (myObj instanceof SomeObject) { }
    // Do anything with myObj here
 }
 
  if (myObj != null) { 
   // Do anything with myObj here
    if (myObj instanceof SomeObject) { }
 }

```
Suppose I need to do some sort of computation with `myObj` (before or after instanceof code) and this computations require it to be non-null so this shouldn't be a redundant condition right? 
if yes, how the check will know if  `myObj` is used somewhere else in the first `if` block so, it doesn't violate it? IMO, we should support this check  when the nullity check is in the same expression with instanceof only. 

for the check name, yes we don't use unneeded I am ok with No or Unnecessary.  but for the word Check in the middle of the check name I can't think of good name without it . `UnnecessarNullCondition` ?

**Comment by nrmancuso** [comment]
I have no problem with having "Check" in the class name twice, since we do not expose this to end users.

I would prefer `NoNullCheckWithInstanceof`
1. `instanceof` is the operator, we should not do `InstanceOf` which by camelcase would imply it is two words (it is not)
2. This makes the check's intention clear in the shortest possible name

I would like the check's message to be like:
`Unnecessary nullity check with instanceof operator in conditional expression.`

I would like to create this check only handle simple cases at first, within the same expression/condition. No nested if's etc. We don't need to figure everything out at once, and it is better to get the check into the hands of users and get real feedback to help guide the evolution of this check.

**Comment by rdiachenko** [comment]
Few more name options: `AvoidNullWithInstanceofCheck`, `NoNullWithInstanceofCheck`.

I lean more towards `NoNullWithInstanceofCheck`. In both cases, having ‘Check’ at the end signifies that it is a condition being checked, as well as indicating that it is a check in Checkstyle terms.

**Comment by mahfouz72** [comment]
@rnveach @nrmancuso @romani @rdiachenko  Please vote
1. NoNullCheckWithInstanceof (🚀  )
2.  UnnecessaryNullCheckWithInstanceof( 👀 )
3.  NoNullWithInstanceof    (👍  )
4.  UnnecessaryNullWithInstanceof (👎 )

**Comment by nrmancuso** [comment]
Ok, looks like `UnnecessaryNullCheckWithInstanceof` is the winner, I am approving this issue based on this. Looks like the description has already been updated.

**Comment by romani** [comment]
@Anmol202005 , please help us to make this module

**Comment by Anmol202005** [comment]
@romani 
```java
if (obj != null && a==b && obj instanceof String) {   
            System.out.println("string");
        }
```
this should be included as a violation or not !

**Comment by mahfouz72** [comment]
Yes, `obj != null ` is redundant.
